Changes since 0.4.4

xcompass opened the pull request:
This is a PR for EdX review and PR is based from version 0.4.4 which is the current version in EdX. All changes are already in master. The changes include:
* Updated UX based on feedbacks from the student usability testing
* Updated UX and accessibility based on feedbacks from EdX
* Added serialisation support for course export 
* Added EdX analytic event (https://openedx.atlassian.net/wiki/display/AN/Peer+Instruction+XBlock+Event+Design)
* Disabled student empty rationale submission 
* Fixed a bug where a 500 error is generated when removing an option from list under certain condition
* Fixed a bug where the responses are missing after going to a different page

sarina commented on ubcpi/static/css/ubcpi.css:
```diff
@@ -349,17 +479,66 @@ textarea.pi-options {
+    margin-bottom: 5px;
+}
+
+.show-correct {
```
I'm worried the name of this selector is too generic might clobber other names? (same with `.show-incorrect` )

richardtape replied:
Darn it, good catch. I've just made a PR to our own repo to fix up these potential clashes. We'll get it into this merged PR asap. Thanks for pointing it out.

sarina replied:
It may be worth just prefacing your selectors with "ubcpi-" to guarantee
uniqueness.

On Tue, Dec 15, 2015 at 1:36 PM, Richard Tape <notifications@github.com>
wrote:

> In ubcpi/static/css/ubcpi.css
> <https://github.com/ubc/ubcpi/pull/55#discussion_r47677333>:
>
> > @@ -349,17 +479,66 @@ textarea.pi-options {
> >  }
> >
> >  .correct-answer-option {
> > -    margin-bottom: 20px;
> > +    margin-top: 20px;
> > +    margin-bottom: 5px;
> > +}
> > +
> > +.show-correct {
>
> Darn it, good catch. I've just made a PR to our own repo to fix up these
> potential clashes. We'll get it into this merged PR asap. Thanks for
> pointing it out.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ubc/ubcpi/pull/55/files#r47677333>.
>

sarina commented:
At a high level these changes look good to me. I've asked @nedbat to also take a look over it, and I'm handing off the analytics events to the analytics team (discussed on wiki page). I am building a sandbox and will ask Alison to take a look at it as well.

nedbat commented on requirements/dev.txt:
```diff
@@ -11,3 +11,5 @@ Werkzeug==0.10.4
 # required by edx-submission, other package will install 2015 version otherwise
 pytz==2012h
+
+-e git://github.com/edx/xblock-utils.git#egg=xblock-utils
```
helpful tip: It's better to not use -e installs if you don't need them.

xcompass replied:
Thanks. Got catch!

nedbat commented on ubcpi/serialize.py:
```diff
@@ -242,7 +244,103 @@ def parse_from_xml(root):
+
+            if hasattr(block, 'correct_rationale'):
+                rationale = etree.SubElement(option, 'rationale')
+                rationale.text = unicode(block.correct_rationale['text'])
```
The use of `unicode()` with no encoding argument concerns me: if the argument is an ascii string, it will work fine, and if it is not a string at all, it will work fine, but if the argument is a string with non-ascii characters, then it will raise an exception.  Please be sure to test thoroughly with non-ascii text in all fields to be sure that this is working properly.

xcompass replied:
It seems the texts in block are already encoded. If I add the encode='utf-8', I got the erorr "decoding Unicode is not supported".
Though I added more tests for non-ascii strings as you suggested.

nedbat replied:
If these values are already unicode strings, then let's drop the `unicode()` function call.

xcompass replied:
OK. Just left some of them that are used to convert int to string.

nedbat commented on ubcpi/serialize.py:
```diff
@@ -242,7 +244,103 @@ def parse_from_xml(root):
+    Returns:
+        None
+    """
+    for index, seed_dict in enumerate(block.seeds):
```
You aren't using `index`, so there's no need for it here.  You can simplify this to: `for seed_dict in block.seeds:`

nedbat commented on ubcpi/serialize.py:
```diff
@@ -242,7 +244,103 @@ def parse_from_xml(root):
+
+    if block.rationale_size is not None:
+        if block.rationale_size.get('min'):
+            root.set('rationale_size_min', unicode(block.rationale_size.get('min', 1)))
```
There's no need to default 'min' to 1 with .get(), since you've already checked it in the previous line.  Or is there something wrong with the logic here?

nedbat commented on ubcpi/serialize.py:
```diff
@@ -242,7 +244,103 @@ def parse_from_xml(root):
+        if block.algo.get('name'):
+            root.set('algorithm', unicode(block.algo.get('name', 'simple')))
+        if block.algo.get('num_responses'):
+            root.set('num_responses', unicode(block.algo.get('num_responses', '#')))
```
Two more uses of `d.get()` where the key has already been checked.

nedbat commented on ubcpi/serialize.py:
```diff
@@ -242,7 +244,103 @@ def parse_from_xml(root):
+        seed.text = unicode(seed_dict.get('rationale', ''))
+
+
+def serialize_to_xml(block, root):
```
It might be easier to follow the code if you used a uniform signature.  Sometimes it's `(data, xml_element)`, and sometimes it's `(xml_element, data)`.

nedbat commented:
My only two concerns are: 1) testing with non-ascii text, and 2) double-check the defaulting logic in the serialization.

xcompass commented:
@nedbat Thanks for all your comments. Added more tests for non-ascii text and removed defaults as those fields will have default value from the xblock. And some tests already there for testing those logics (those "minimal" test cases).

sarina commented:
@nedbat any remaining comments?

nedbat commented:
:+1: looks good, thanks :)

jbarciauskas commented:
@nedbat what's the next step here? Seems like we need another OSPR ticket to track this, or link this to https://openedx.atlassian.net/browse/OSPR-850 somehow

xcompass commented:
@jbarciauskas we are still waiting for Ned's response for this PR: https://github.com/ubc/ubcpi/pull/54 to decided it we can include scoring this this release.

xcompass commented:
@jbarciauskas I merged that scoring piece in. Do we need another round review or we are good? I can update the tag in PR for edx-platform to the latest version.

xcompass commented:
@jbarciauskas Any update? Thanks

jbarciauskas commented:
Hi @xcompass, sorry for the delay - I was out for the holidays, I meant to include a comment here to that effect. If @nedbat is good with this then I think we're ready to go.

xcompass commented:
@nedbat just following this up. Any concerns remaining?

nedbat commented:
:+1:

nedbat commented:
I've merged https://github.com/edx/edx-platform/pull/10937, so you can go ahead and merge this. :)
